### PR TITLE
feat: adding aws instance types

### PR DESCRIPTION
**Pull request by trentontrees**
Added support for:
- t2.nano
- t2.xlarge
- t2.2xlarge

- t3.nano
- t3.micro
- t3.small
- t3.medium
- t3.large
- t3.xlarge
- t3.2xlarge

- m4.16xlarge

AWS Instance Type reference page:
https://aws.amazon.com/ec2/instance-types/

### CODE REVIEW

**Comment by trentontrees** [comment]
image

Verified via local IT

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 1]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
             .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
    }
 
+   /**
+    * @see InstanceType#T2_XLARGE
+    */
+   public static EC2HardwareBuilder t2_xlarge() {
+      return new EC2HardwareBuilder(InstanceType.T2_XLARGE).t2()
+              .ram(16384)
+              .processors(ImmutableList.of(new Processor(2.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (vapgar200) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:440]: assuming the first param of the Processor constructor is the number of vCPUs, then this should be 4, not 2 for a t2.xlarge

Comment (VRanga000) [reply apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:440]: good catch, yup ```

t2.xlarge | 4 | 16
-- | -- | --
t2.2xlarge | 8 | 32
```

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 2]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T2_2XLARGE
+    */
+   public static EC2HardwareBuilder t2_2xlarge() {
+      return new EC2HardwareBuilder(InstanceType.T2_2XLARGE).t2()
+              .ram(32768)
+              .processors(ImmutableList.of(new Processor(4.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:449]: number o0f vcpus here should be 8: 
```

t2.2xlarge | 8
-- | --
```

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 3]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_NANO
+    */
+   public static EC2HardwareBuilder t3_nano() {
+      return new EC2HardwareBuilder(InstanceType.T3_NANO).t3()
+              .ram(512)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:458]: vcpus=2

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 4]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_MICRO
+    */
+   public static EC2HardwareBuilder t3_micro() {
+      return new EC2HardwareBuilder(InstanceType.T3_MICRO).t3()
+              .ram(1024)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:467]: vpcus=2

**Comment by VRanga000** [comment]
General comment on the instance specs being off - not sure exactly how these matter (beyond "documentation") - because the instance type is the only thing ultimately passed on to aws APIs (as a string/enum). But I guess since they are in there, we have to be consistent/compliant.

File: providers/dynect/src/test/resources/logback.xml [hunk 5]
```diff
@@ -1,38 +1,38 @@
...
-        <appender-ref ref="WIREFILE" />
-    </logger>
-
-    <logger name="jclouds.headers">
-        <level value="DEBUG" />
-        <appender-ref ref="WIREFILE" />
-    </logger>
-
-</configuration>
+<?xml version="1.0"?>
```
Comment (VRanga000) [inline providers/dynect/src/test/resources/logback.xml:1]: are there just whitespace changes in this file?

File: providers/ultradns-ws/src/test/resources/logback.xml [hunk 6]
```diff
@@ -1,38 +1,38 @@
...
-        <appender-ref ref="WIREFILE" />
-    </logger>
-
-    <logger name="jclouds.headers">
-        <level value="DEBUG" />
-        <appender-ref ref="WIREFILE" />
-    </logger>
-
-</configuration>
+<?xml version="1.0"?>
```
Comment (VRanga000) [inline providers/ultradns-ws/src/test/resources/logback.xml:1]: ditto

File: providers/aws-ec2/src/main/java/org/jclouds/aws/ec2/compute/suppliers/AWSEC2HardwareSupplier.java [hunk 7]
```diff
@@ -159,6 +180,7 @@ public AWSEC2HardwareSupplier(@ClusterCompute Set<String> ccAmis) {
       sizes.add(m4_2xlarge().build());
       sizes.add(m4_4xlarge().build());
       sizes.add(m4_10xlarge().build());
+      sizes.add(m4_16xlarge().build());
```
Comment (VRanga000) [inline providers/aws-ec2/src/main/java/org/jclouds/aws/ec2/compute/suppliers/AWSEC2HardwareSupplier.java:183]: Thanks for catching this missed subtype :)

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 8]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_MEDIUM
+    */
+   public static EC2HardwareBuilder t3_medium() {
+      return new EC2HardwareBuilder(InstanceType.T3_MEDIUM).t3()
+              .ram(4096)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:485]: vcpus=2

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 9]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_SMALL
+    */
+   public static EC2HardwareBuilder t3_small() {
+      return new EC2HardwareBuilder(InstanceType.T3_SMALL).t3()
+              .ram(2048)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:476]: vcpus=2

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 10]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_LARGE
+    */
+   public static EC2HardwareBuilder t3_large() {
+      return new EC2HardwareBuilder(InstanceType.T3_LARGE).t3()
+              .ram(8192)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:494]: vcpus=2

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 11]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_XLARGE
+    */
+   public static EC2HardwareBuilder t3_xlarge() {
+      return new EC2HardwareBuilder(InstanceType.T3_XLARGE).t3()
+              .ram(16384)
+              .processors(ImmutableList.of(new Processor(2.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:503]: vcpus=4

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 12]
```diff
@@ -416,6 +431,89 @@ public static EC2HardwareBuilder t2_large() {
...
+   }
+
+   /**
+    * @see InstanceType#T3_2XLARGE
+    */
+   public static EC2HardwareBuilder t3_2xlarge() {
+      return new EC2HardwareBuilder(InstanceType.T3_2XLARGE).t3()
+              .ram(32768)
+              .processors(ImmutableList.of(new Processor(4.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (VRanga000) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java:512]: vcpus=8

**Comment by VRanga000** [comment]
> image
> 
> Verified via local IT

If you have the IT tests, make sense to PR them in gumby-lib? If not I can..

**Pull request closed without merging**